dm/syncer/safe_mode.go

```diff
@@ -43,35 +60,90 @@ func (s *Syncer) enableSafeModeInitializationPhase(tctx *tcontext.Context) {
+
+	err = dumpling.ParseExtraArgs(&logger, exportCfg, strings.Fields(s.cfg.MydumperConfig.ExtraArgs))
+	if err != nil {
+		return
+	}
+	fresh, err := s.IsFreshTask(tctx.Ctx)
+	if err != nil {
+		return
+	}
+	var initPhaseSeconds string
+	// SafeModeDuration doesn't work when task start first with consistency is none
+	if !(exportCfg.Consistency == export.ConsistencyTypeNone && fresh) {
+		initPhaseSeconds = s.cfg.SafeModeDuration
+	}
+
+	failpoint.Inject("SafeModeInitPhaseSeconds", func(val failpoint.Value) {
+		initPhaseSeconds = val.(string)
+		s.tctx.L().Info("set initPhaseSeconds", zap.String("failpoint", "SafeModeInitPhaseSeconds"), zap.String("value", initPhaseSeconds))
+	})
+	failpoint.Inject("SafeModeDurationSet", func(val failpoint.Value) {
+		initPhaseSeconds = val.(string)
+		s.tctx.L().Info("set initPhaseSeconds", zap.String("failpoint", "SafeModeDurationSet"), zap.String("value", initPhaseSeconds))
+	})
+	if initPhaseSeconds == "" {
+		duration = time.Second * time.Duration(2*s.cfg.CheckpointFlushInterval)
+	} else {
+		duration, err = time.ParseDuration(initPhaseSeconds)
+		if err != nil {
+			s.tctx.L().Error("enable safe-mode failed due to duration parse failed", zap.String("duration", initPhaseSeconds))
+			return
+		}
+	}
+	exitPoint := s.checkpoint.SafeModeExitPoint()
+	beginLocation := s.checkpoint.GlobalPoint()
+	failpoint.Inject("SafeModeDurationSetBeginLoc", func() {
+		beginLocation = binlog.MustZeroLocation(mysql.MySQLFlavor)
+	})
+	if exitPoint != nil {
+		s.tctx.L().Info("compare exitPoint and beginLocation", zap.Stringer("exitPoint", exitPoint), zap.Stringer("beginLocation", beginLocation))
+		if binlog.CompareLocation(*exitPoint, beginLocation, s.cfg.EnableGTID) == 0 {
+			s.tctx.L().Info("exitPoint equal to beginLocation, so disable the safe mode")
+			s.checkpoint.SaveSafeModeExitPoint(nil)
+			return
+		}
+
 		//nolint:errcheck
 		s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
-		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *s.checkpoint.SafeModeExitPoint()))
+		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *exitPoint))
 	} else {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
-		go func() {
-			defer func() {
-				err := s.safeMode.Add(tctx, -1)
-				if err != nil {
-					// send error to the fatal chan to interrupt the process
-					s.runFatalChan <- unit.NewProcessError(err)
-				}
-				if !s.safeMode.Enable() {
-					s.tctx.L().Info("disable safe-mode after task initialization finished")
+		s.tctx.L().Info("enable safe-mode because of task initialization", zap.Duration("duration", duration))
+
+		if int64(duration) > 0 {
+			//nolint:errcheck
+			s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
+			go func() {
+				defer func() {
+					err = s.safeMode.Add(tctx, -1)
```

Comment:
this `err` may race with the accessing from outer goroutine.

Reply:
Good catch! I will update it.